port/linux/omrvmem.c

```diff
@@ -829,10 +843,14 @@ omrvmem_reserve_memory_ex(struct OMRPortLibrary *portLibrary, struct J9PortVmemI
 				uintptr_t defaultPageSize = PPG_vmem_pageSize[0];
 				uintptr_t alignmentInBytes = OMR_MAX(defaultPageSize, params->alignmentInBytes);
 				uintptr_t minimumGranule = OMR_MIN(defaultPageSize, params->alignmentInBytes);
+#if defined(MAP_HUGETLB)
+				/* Get rid of large pages and retry with default pages */
+				params->mode &= ~OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES;
+#endif /* MAP_HUGETLB */
```

Comment:
I do not insist, but set flag before and clear it after looks a little bit ugly, is not it?
I can see two solutions I am not sure it is better:
- Cache params->mode before flag is set. Used cached value here (assumption here is nobody set extra flags between - not so good for me)...
- or just add flag to first call of `getMemoryInRangeWithMmap()` above explicitly. Flag is set under `#ifdef` so use different variable. Should this flag be passed to `reserveMemoryWithShmat()` call?
```
    uintptr_t extraModeFlags = 0;
.....
    extraModeFlags |= OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES;
.....
		/* If MAP_HUGETLB is available we can reserve large pages with mmap */
  			if (0 != (params->mode & OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES)) {
  				memoryPointer = getMemoryInRangeWithMmap(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, largePageAlignmentInBytes, params->options, params->mode | extraModeFlags, PPG_vmem_pageSize[1]);
  			} else {
  				memoryPointer = reserveMemoryWithShmat(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, params->pageSize, largePageAlignmentInBytes, params->options, params->mode);
  			}
```
What do you think?

Reply:
Yeah, I think your second suggestion might be better because we don't need to pass this flag to `reserveMemoryWithShmat()`, even if we do, it would just ignore it. But we would still need to check it against `portLibrary->portGlobals->huge_pages_mmap_enabled`. What if we do something like?
```
uintptr_t mmapMode = PPG_huge_pages_mmap_enabled ? (params->mode | OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES) : params->mode;
...
/* If MAP_HUGETLB is available we can reserve large pages with mmap */
  			if (0 != (params->mode & OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES)) {
  				memoryPointer = getMemoryInRangeWithMmap(portLibrary, identifier, category, params->byteAmount, params->startAddress, params->endAddress, largePageAlignmentInBytes, params->options, mmapMode, PPG_vmem_pageSize[1]);
  			} else {
```
That way we wouldn't need to clear flag

Reply:
Yes, it is (about) what I had in mind. I just did not show check for `PPG_huge_pages_mmap_enabled` in my example.
Your suggestion is an equivalent. There is just stylish thing. In my case `extraModeFlags` added explicitly at the place it should be. In your version code reader should recognize that `getMemoryInRangeWithMmap()` is called with `mmapMode` but `reserveMemoryWithShmat()` still use `params->mode`. As I said it is stylish and I agree with your suggestion

Reply:
I would just avoid to use an arithmetic `if`: `if (0 != PPG_huge_pages_mmap_enabled) {`

Reply:
Just want to make sure I understood. Rather use `if (0 != PPG_huge_pages_mmap_enabled) { `over `if (PPG_huge_pages_mmap_enabled) {` or the other way around?

Reply:
sorry, was not clear... `if (PPG_huge_pages_mmap_enabled) {` is arithmetic because type of `PPG_huge_pages_mmap_enabled` is `uintptr_t`. Using `if (0 != PPG_huge_pages_mmap_enabled)` is preferable because condition is logical

Reply:
I see, makes sense, thanks!